VisualiserTool : Vector visualiser

ericmehl opened the pull request:
This adds a new visualiser mode for showing vector values.

I didn't add a dedicated mode for vector visualisation because I don't know that it makes much sense to visualise anything except V3f data. That's covered by the `Auto` mode. But I'm open to other ideas if we think it would be good to have such a mode.

I also added recognition of a new metadata entry for `ColorPlugValueWidget` - `colorPlugValueWidget:colorChooserButtonVisible`. This is for hiding the button to display the inline color picker on the `vectorColor` plug. Allowing that to be shown on the viewer tool shifts all the tools around and is fairly messy. I didn't see an entry for the already existing `colorPlugValueWidget:colorChooserVisible` metadata in the change log or documentation, so I'm following that lead here and treating my new metadata as "private".

There's one outstanding issue that I haven't sorted out yet. The test at https://github.com/ericmehl/gaffer/blob/01e23d87960cb22df228353498c51110ad85ff31/src/GafferSceneUI/VisualiserTool.cpp#L2373-L2379 prevents doing the mesh evaluation to get interpolated values for the variable when we're doing vertex label visualisation and the value can't be acquired (for example if the variable does not use vertex interpolation).

I'm piggy-backing on the vertex label code to figure out the closest point. But this introduces another case that needs to prevent doing the interpolation. That is, if we are doing vector visualisation and the value doesn't have vertex interpolation. I'm checking the cursor value to see if it is `std::monostate`, i.e.  not set to anything meaningful, to indicate that finding the closest point wasn't done, so we should interpolate the value. But now there are two reasons it might be `monostate` - we're meant to be drawing interpolated values, or we want to draw per-vertex values but can't.

I went down a few avenues to fix that, but none were feeling all that satisfactory, so I'm all ears on a better solution.

### Checklist ###

- [X] I have read the [contribution guidelines](https://github.com/GafferHQ/gaffer/blob/main/CONTRIBUTING.md).
- [X] I have updated the documentation, if applicable.
- [X] I have tested my change(s) in the test suite, and added new test cases where necessary.
- [X] My code follows the Gaffer project's prevailing coding style and conventions.

ericmehl commented:
I think it might be possible to improve the way we figure out which visualiser to use for each object, and maybe improve how we set the cursor value as well. Currently we call each visualiser with the mode and let them decide if they want to draw. That has led to the code determining which visualiser is active being spread out across multiple places. It also means a lot of the logic for that is "if I shouldn't render, continue on with the next object", instead of "I should render" which I've found to be awkward to reason through at times (logically positive is easier than negative to me).

Maybe this would be an improvement: loop through all the selected objects in `renderLayer()`, get the object data, check for variables and variable type and all that jazz. In that loop we build a list of paths to visualise for each visualiser. Then we pass that list to the visualisers and say "visualise this, we've already sorted out you can and should do it".

That way we're still setting up the OpenGL state for each visualiser only once. But the logic of which visualiser to use per object is all in one place, and we would also reduce the duplicate code for getting variables and checking their existence and type. And I _think_ there would be a way to sort out who should be responsible for setting the cursor value (currently nearest vertex vs. interpolated mesh sample).

But I haven't gone through this idea in great detail so maybe there would be a pitfall in there somewhere. Worth investigating before reviewing this in detail?

murraystevenson reviewed (changes requested):
Thanks Eric! A few nitpicks inline and some thoughts below.

I like the updated presentation of uniform data, though it's unfortunate that we miss out on the positive feedback received from displaying labels on hover. I felt a bit blind trying to drag and drop a uniform N value. Could be worth reusing the resample P to uniform trick on vertex labels like you mentioned this morning. Happy for that to be a separate PR.

I also found it a bit confusing that the `opacity` plug doesn't do anything in vector mode. Maybe this just needs clarification on the plug description.

> I didn't add a dedicated mode for vector visualisation because I don't know that it makes much sense to visualise anything except V3f data. That's covered by the Auto mode. But I'm open to other ideas if we think it would be good to have such a mode.

I think this is fine for the moment, though in the future we may want other ways of colouring the vector visualisations, so it might make sense to have those as additional modes?

> I think it might be possible to improve the way we figure out which visualiser to use for each object, and maybe improve how we set the cursor value as well.

Yeah, that might be worth tackling as a refactor once we have the functionality in place. I think there's also some stuff to be improved with the cursor values when moving the camera while a value is drawn. Right now we get a fair amount of disconnect once the camera starts moving, while it may be better for the current value to stick to help you navigate the camera closer to your point of interest...

ericmehl commented on python/GafferSceneUI/VisualiserToolUI.py:
```diff
@@ -158,6 +159,34 @@
+			""",
+
+			"toolbarLayout:section", "Bottom",
+			"toolbarLayout:width", 100,
```
That does work out better being a little more compact. Fixed in https://github.com/GafferHQ/gaffer/pull/6276/commits/58e3dc4069b6a4b35533681c7266f687a6ef9dbe

ericmehl commented on python/GafferSceneUI/VisualiserToolUI.py:
```diff
@@ -158,6 +159,34 @@
+
+			"description",
+			"""
+			The color to use for drawing vectors.
```
Fixed in https://github.com/GafferHQ/gaffer/pull/6276/commits/ad144e67d9098638c49e949969706ab2008a726c

ericmehl commented on src/GafferSceneUI/VisualiserTool.cpp:
```diff
@@ -1351,6 +1505,258 @@ class VisualiserGadget : public Gadget
+					case GeometricData::Interpretation::Normal :
+						vDataProgram = m_vectorShaderBivector->program();
+						break;
+					default : vDataProgram = m_vectorShaderVector->program();
```
Fixed in https://github.com/GafferHQ/gaffer/pull/6276/commits/b3aaf2984dc70105e0aa0557681f9d9891e14445

ericmehl commented on python/GafferSceneUI/VisualiserToolUI.py:
```diff
@@ -158,6 +159,34 @@
+
+		],
+
+		"vectorColor" : [
```
You're right, that does diverge noticeably between the vector lines and the color swatch. In https://github.com/GafferHQ/gaffer/pull/6276/commits/8ef00688283cd4e5a18f285dcd805088897029bc I added a simple custom widget to set the widget to a identity transform.

I also needed to pipe through the display transform from the swatch widget to the dialog you get when clicking on the swatch. That dialog was parented to the script window, so it inherited the global display transform instead of the swatch's transform.

ericmehl replied:
Does the change in the dialog deserve a changelog entry? I don't know anywhere else this is an issue, at least in stock Gaffer. Maybe I'm missing something, or it's worth mentioning if someone has relevant extensions?

ericmehl replied:
Added in https://github.com/GafferHQ/gaffer/pull/6276/commits/069878d6795430c0407d6e18d064233c41e811b0

ericmehl commented on src/GafferSceneUI/VisualiserTool.cpp:
```diff
@@ -1290,9 +1428,25 @@ class VisualiserGadget : public Gadget
+										else
+										{
+											cursorVertexValueTextScale = 2.f;
+										}
```
Yes, that's definitely able to be moved up and out of the loop. It's more logically grouped with the `vData` handling in https://github.com/GafferHQ/gaffer/pull/6276/commits/16d93800c4400678d4ead3663634908dd118f154

ericmehl commented:
Thanks Murray! Those are all nice additions and fixes spotted.

> I also found it a bit confusing that the `opacity` plug doesn't do anything in vector mode. Maybe this just needs clarification on the plug description.

I added it in https://github.com/GafferHQ/gaffer/pull/6276/commits/e6cec103921257ea6f6997aef3deb7803b6d54a8. It meant changing the blending mode, which I can't see any differences when comparing the drawing before and after so I think it's good.

> Could be worth reusing the resample P to uniform trick on vertex labels like you mentioned this morning

That's pretty easy with the groundwork in place for the uniform vector visualisation, so I added it in https://github.com/GafferHQ/gaffer/pull/6276/commits/73dde9b6ce9513346cabf485b91ab607ad4ab950.

Comments are now addressed and ready for another look!

ericmehl commented on src/GafferSceneUI/VisualiserTool.cpp:
```diff
@@ -1351,6 +1531,259 @@ class VisualiserGadget : public Gadget
+			}
+
+			const GLboolean blendEnabled = glIsEnabled( GL_BLEND );
+			if( blendEnabled )
```
Yes we certainly do, done in https://github.com/GafferHQ/gaffer/pull/6276/commits/09d01fdd4185a80ab3efd6e62b08972f5deef4e0

ericmehl commented:
Hmmm, the inconsistent uniform labels might be a tricky one. You're right that they are omitted as a result of a depth test. For the vertex label display this piggybacks on, we first draw the points to a buffer (including a depth test) to see which are visible. So if the face is not quite planar, the average position of the vertices might end up behind the face.

I'm not seeing a way to improve the uniform data drawing with this technique. Maybe we do a follow-up PR where we take the vertex visibility concept and extend it to work with full faces? And drop the uniform labels for this PR?

murraystevenson commented:
> Maybe we do a follow-up PR where we take the vertex visibility concept and extend it to work with full faces? And drop the uniform labels for this PR?

Sounds good to me, it makes sense not to hold this PR up.

ericmehl commented:
Great, I squashed the fixups down and dropped the commit adding the uniform labels. There's one bit of that commit I needed to keep to make the labels centered which is in https://github.com/GafferHQ/gaffer/pull/6276/commits/9acf37c4edf59b5bf2ca11a3da52c27b064302fd.

Probably worth taking it for another spin to make sure it all feels good for now, and if all is well hopefully we're ready for merging!

murraystevenson commented:
Thanks Eric! Looks good to squash and merge.

ericmehl commented:
Squashed and merging...
